hedera-node/src/main/java/com/hedera/services/store/tokens/unique/UniqueTokenStore.java

```diff
@@ -63,34 +76,117 @@ public UniqueTokenStore(final EntityIdSource ids,
+		this.lastMintedSerialNumbers = new ArrayList<>();
+	}
+
+	@Override
+	public List<Long> popLastMintedSerialNumbers() {
+		var copy = new ArrayList<>(lastMintedSerialNumbers);
+		lastMintedSerialNumbers.clear();
+		return copy;
+	}
+
+	@Override
+	public ResponseCodeEnum wipe(final AccountID aId, final TokenID tId, final long wipingAmount, final boolean skipKeyCheck) {
+		return null;
+	}
+
+	@Override
+	public ResponseCodeEnum mintProvisional(final TokenMintTransactionBody txBody, final RichInstant creationTime) {
+		return tokenSanityCheck(txBody.getToken(), merkleToken -> {
+			if (!merkleToken.hasSupplyKey()) {
+				return TOKEN_HAS_NO_SUPPLY_KEY;
+			}
+			final var eId = EntityId.fromGrpcTokenId(txBody.getToken());
+			final var owner = merkleToken.treasury();
+			final var metadataList = txBody.getMetadataList();
+			metadataList.stream().map(el -> {
+				String metaAsStr = el.toStringUtf8();
+				final long serialNum = merkleToken.getCurrentSerialNum(); // create them without incrementing serialNum
+				final var nftId = new MerkleUniqueTokenId(eId, serialNum);
+				final var nft = new MerkleUniqueToken(owner, metaAsStr, creationTime);
+				return Pair.of(nftId, nft);
+			}).forEach(provisionalUniqueTokens::add);
+			return OK;
+		});
+	}
+
+	@Override
+	public ResponseCodeEnum mint(final TokenMintTransactionBody txBody, final RichInstant creationTime) {
+		var res = mintProvisional(txBody, creationTime);
+		if (!res.equals(OK)) {
+			clearProvisional();
+			return res;
+		}
+
+		res = commitProvisional();
+		if (!res.equals(OK)) {
+			clearProvisional();
+			return res;
+		}
+		clearProvisional();
+		return OK;
 	}
 
 	@Override
-	public ResponseCodeEnum mint(final TokenID tId, final String memo, final RichInstant creationTime) {
+	public ResponseCodeEnum commitProvisional() {
+		final AtomicBoolean didAnyFail = new AtomicBoolean(false);
+		if (!checkProvisional()) {
+			return INVALID_TRANSACTION_BODY;
+		} else {
+			provisionalUniqueTokens.forEach(provisionalTokenPair -> {
+				TokenID tokenId = provisionalTokenPair.getKey().tokenId().toGrpcTokenId();
+				String memo = provisionalTokenPair.getValue().getMemo();
+				RichInstant creationTime = provisionalTokenPair.getValue().getCreationTime();
+				var res = mintOne(tokenId, memo, creationTime);
+				// if native minting fails for some reason, revert changes
+				// serial num will not be incremented if we go here
+				if (!res.equals(OK)) {
+					var token = get(tokenId);
+					revertAdjustmentOfToken(token, tokenId);
+					// revert from map (remove)
+					uniqueTokenSupplier.get().remove(new MerkleUniqueTokenId(EntityId.fromGrpcTokenId(tokenId), token.getCurrentSerialNum()));
+					didAnyFail.set(true);
+				}
+			});
+			return didAnyFail.get() ? FAIL_INVALID : OK;
+		}
+	}
+
+	private ResponseCodeEnum mintOne(final TokenID tId, final String memo, final RichInstant creationTime) {
 		return tokenSanityCheck(tId, (merkleToken -> {
```

Comment:
You do not need the tokenSanityCheck and supply key check performed on every NFT minted. Its done once at `mint`